File: IAIso-v5.0/skills/loader/loader.py
"""
Minimal IAIso skill loader.

Reads ./skills/<name>/SKILL.md files and exposes them as a
queryable registry. No external dependencies.

Usage:
    from skills.loader.loader import SkillRegistry
    reg = SkillRegistry.load("./skills")
    skill = reg["iaiso-runtime-governed-agent"]
    print(skill.body)
"""

from __future__ import annotations
import os
import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Iterator


_FRONTMATTER_RE = re.compile(r"^---\n(.*?)\n---\n(.*)$", re.DOTALL)


@dataclass
class Skill:
    name: str
    description: str = ""
    version: str = ""
    tier: str = ""
    category: str = ""
    framework: str = ""
    license: str = ""
    body: str = ""
    metadata: dict = field(default_factory=dict)
    path: str = ""

    @classmethod
    def from_file(cls, path: str | os.PathLike) -> "Skill":
        text = Path(path).read_text(encoding="utf-8")
        m = _FRONTMATTER_RE.match(text)
        if not m:
            raise ValueError(f"{path}: missing or malformed frontmatter")
        fm_text, body = m.group(1), m.group(2)
        meta: dict = {}
        for line in fm_text.splitlines():
            if not line.strip() or ":" not in line:
                continue
            key, _, value = line.partition(":")
            meta[key.strip()] = value.strip().strip('"').strip("'")
        return cls(
            name=meta.pop("name", Path(path).parent.name),
            description=meta.pop("description", ""),
            version=meta.pop("version", ""),
            tier=meta.pop("tier", ""),
            category=meta.pop("category", ""),
            framework=meta.pop("framework", ""),
            license=meta.pop("license", ""),
            body=body.lstrip(),
            metadata=meta,
            path=str(path),
        )


class SkillRegistry:
    def __init__(self, skills: list[Skill]):
        self._by_name: dict[str, Skill] = {s.name: s for s in skills}

    @classmethod
    def load(cls, root: str | os.PathLike) -> "SkillRegistry":
        root_p = Path(root)
        skills: list[Skill] = []
        for skill_md in root_p.glob("*/SKILL.md"):
            try:
                skills.append(Skill.from_file(skill_md))
            except ValueError as e:
                # Skip malformed; surface as a warning in real use.
                print(f"warn: {e}")
        return cls(skills)

    def __getitem__(self, name: str) -> Skill:
        return self._by_name[name]

    def get(self, name: str, default=None):
        return self._by_name.get(name, default)

    def __contains__(self, name: str) -> bool:
        return name in self._by_name

    def __iter__(self) -> Iterator[Skill]:
        return iter(self._by_name.values())

    def __len__(self) -> int:
        return len(self._by_name)

    def tier(self, t: str) -> list[Skill]:
        return [s for s in self if s.tier == t]

    def category(self, c: str) -> list[Skill]:
        return [s for s in self if s.category == c]

    def names(self) -> list[str]:
        return sorted(self._by_name)


if __name__ == "__main__":
    import sys
    root = sys.argv[1] if len(sys.argv) > 1 else "."
    reg = SkillRegistry.load(root)
    print(f"loaded {len(reg)} skills from {root}")
    by_tier = {}
    for s in reg:
        by_tier.setdefault(s.tier, 0)
        by_tier[s.tier] += 1
    for t in sorted(by_tier):
        print(f"  {t}: {by_tier[t]}")

File: IAIso-v5.0/skills/loader/loader.ts
/**
 * Minimal IAIso skill loader (TypeScript / Node).
 *
 * Reads ./skills/<name>/SKILL.md files and exposes them as a
 * queryable registry. Zero runtime dependencies beyond Node's
 * built-in fs/path.
 *
 * Usage:
 *   import { SkillRegistry } from "./skills/loader/loader";
 *   const reg = await SkillRegistry.load("./skills");
 *   const skill = reg.get("iaiso-runtime-governed-agent");
 *   console.log(skill?.body);
 */

import * as fs from "node:fs/promises";
import * as path from "node:path";

export interface Skill {
  name: string;
  description: string;
  version: string;
  tier: string;
  category: string;
  framework: string;
  license: string;
  body: string;
  metadata: Record<string, string>;
  path: string;
}

const FM_RE = /^---\n([\s\S]*?)\n---\n([\s\S]*)$/;

function parseSkill(p: string, text: string): Skill {
  const m = text.match(FM_RE);
  if (!m) throw new Error(`${p}: missing or malformed frontmatter`);
  const [, fmText, body] = m;
  const meta: Record<string, string> = {};
  for (const line of fmText.split(/\n/)) {
    const idx = line.indexOf(":");
    if (idx < 0) continue;
    const key = line.slice(0, idx).trim();
    let val = line.slice(idx + 1).trim();
    if ((val.startsWith('"') && val.endsWith('"')) ||
        (val.startsWith("'") && val.endsWith("'"))) {
      val = val.slice(1, -1);
    }
    meta[key] = val;
  }
  const take = (k: string, fallback = ""): string => {
    const v = meta[k] ?? fallback;
    delete meta[k];
    return v;
  };
  return {
    name: take("name", path.basename(path.dirname(p))),
    description: take("description"),
    version: take("version"),
    tier: take("tier"),
    category: take("category"),
    framework: take("framework"),
    license: take("license"),
    body: body.replace(/^\n+/, ""),
    metadata: meta,
    path: p,
  };
}

export class SkillRegistry {
  private byName: Map<string, Skill>;

  constructor(skills: Skill[]) {
    this.byName = new Map(skills.map((s) => [s.name, s]));
  }

  static async load(root: string): Promise<SkillRegistry> {
    const skills: Skill[] = [];
    const entries = await fs.readdir(root, { withFileTypes: true });
    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      const skillPath = path.join(root, entry.name, "SKILL.md");
      try {
        const text = await fs.readFile(skillPath, "utf-8");
        skills.push(parseSkill(skillPath, text));
      } catch (err: any) {
        if (err.code !== "ENOENT") {
          console.warn(`warn: failed to load ${skillPath}: ${err.message}`);
        }
      }
    }
    return new SkillRegistry(skills);
  }

  get(name: string): Skill | undefined { return this.byName.get(name); }
  has(name: string): boolean { return this.byName.has(name); }
  get size(): number { return this.byName.size; }

  *[Symbol.iterator](): IterableIterator<Skill> {
    yield* this.byName.values();
  }

  tier(t: string): Skill[] {
    return [...this].filter((s) => s.tier === t);
  }
  category(c: string): Skill[] {
    return [...this].filter((s) => s.category === c);
  }
  names(): string[] {
    return [...this.byName.keys()].sort();
  }
}
